### PR TITLE
Recommendation framework

### PR DESCRIPTION
**Pull request by xieydd**
#### What type of PR is this?
Feature
#### What this PR does / why we need it:
Define recommendation framework interfaces .
Implement replicas recommendation via framework.
#### Special notes for your reviewer:
@qmhu

### CODE REVIEW

**Comment by xieydd** [comment]
done. pls review this pr @qmhu

File: pkg/recommendation/replicas/fliter.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package replicas
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/gocrane/api/analysis/v1alpha1"
```
Comment (qmhu) [inline pkg/recommendation/replicas/fliter.go:7]: import format

File: pkg/recommendation/replicas/prepare.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	promDataProvider, err := prom.NewProvider(&promConfig)
+	if err != nil {
+		return err
+	}
+	ctx.DataProviders = map[providers.DataSourceType]providers.History{
+		providers.PrometheusDataSource: promDataProvider,
+	}
+
+	// if no data provider config set, use default history data provider
```
Comment (qmhu) [inline pkg/recommendation/replicas/prepare.go:85]: Do we need to keep this block?

File: pkg/recommendation/replicas/recommend.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+	"github.com/montanaflynn/stats"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+)
+
+func (rr *ReplicasRecommender) PreRecommend(ctx *framework.RecommendationContext) error {
+	// we load algorithm config in this phase
+	// TODO(chrisydxie) support configuration
+	config := &config.Config{
```
Comment (qmhu) [inline pkg/recommendation/replicas/recommend.go:24]: What if we have multiple algo config

File: pkg/recommendation/replicas/recommend.go [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+	min := int32(math.Ceil(workloadCpu / (float64(targetUtilization) / 100. * float64(requestTotal) / 1000.)))
+	if min > minReplicas {
+		minReplicas = min
+	}
+
+	return minReplicas, nil
+}
+
+// proposeMaxReplicas use max cpu usage to compare with target pod cpu usage to get the max replicas.
+//func (rr *ReplicasRecommender) proposeMaxReplicas(ctx *framework.RecommendationContext, percentileCpu float64, targetUtilization int32, minReplicas int32) (int32, error) {
```
Comment (qmhu) [inline pkg/recommendation/replicas/recommend.go:295]: Can we remove these unused code?

File: pkg/recommendation/replicas/recommend.go (hunk 3 above)
Comment (xieydd) [reply pkg/recommendation/replicas/recommend.go:24]: Can you describe the specific user case? Do we need hyper-parameter search?

File: pkg/recommendation/replicas/recommend.go (hunk 4 above)
Comment (xieydd) [reply pkg/recommendation/replicas/recommend.go:295]: lgtm

File: pkg/recommendation/replicas/prepare.go (hunk 2 above)
Comment (xieydd) [reply pkg/recommendation/replicas/prepare.go:85]: we does not need this block, i will delete it.

File: pkg/recommendation/replicas/fliter.go (hunk 1 above)
Comment (xieydd) [reply pkg/recommendation/replicas/fliter.go:7]: fixed.

File: pkg/recommendation/recommender/resource/prepare.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+// CheckDataProviders in PrePrepare phase, will create data source provider via your recommendation config.
+func (rr *ResourceRecommender) CheckDataProviders(ctx *framework.RecommendationContext) error {
+	if err := rr.BaseRecommender.CheckDataProviders(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rr *ResourceRecommender) CollectData(ctx *framework.RecommendationContext) error {
```
Comment (xieydd) [inline pkg/recommendation/recommender/resource/prepare.go:16]: resource no need collect data?

File: pkg/recommendation/recommender/resource/recommend.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	predictionapi "github.com/gocrane/api/prediction/v1alpha1"
+
+	"github.com/gocrane/crane/pkg/metricnaming"
+	"github.com/gocrane/crane/pkg/prediction/config"
+	"github.com/gocrane/crane/pkg/recommend/types"
+	"github.com/gocrane/crane/pkg/recommendation/framework"
+	"github.com/gocrane/crane/pkg/utils"
+)
+
+const callerFormat = "RecommendationCaller-%s-%s"
```
Comment (xieydd) [inline pkg/recommendation/recommender/resource/recommend.go:23]: same as hpa recommender, need change to ResourceRecommederCaller-%s-%s.

File: pkg/recommendation/recommender/resource/recommend.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+func (rr *ResourceRecommender) PreRecommend(ctx *framework.RecommendationContext) error {
+	return nil
```
Comment (xieydd) [inline pkg/recommendation/recommender/resource/recommend.go:26]: no need load algoritjm config?

File: pkg/recommendation/recommender/base/prepare.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	promDataProvider, err := prom.NewProvider(&promConfig)
+	if err != nil {
+		return err
+	}
+	ctx.DataProviders = map[providers.DataSourceType]providers.History{
+		providers.PrometheusDataSource: promDataProvider,
+	}
+
+	// will be use in future.
+	// if no data provider config set, use default history data provider
```
[the remark this answers is not in the document]
Comment (qmhu) [reply pkg/recommendation/recommender/base/prepare.go:86]: we can implement it in future.

File: pkg/recommendation/framework/context.go [hunk 9]
```diff
@@ -73,3 +79,33 @@ func NewRecommendationContext(context context.Context, identity analytics.Object
 //		return false
 //	}
 //}
+
+func ToK8SObject(object interface{}, target interface{}) error {
```
[the remark this answers is not in the document]
Comment (qmhu) [reply pkg/recommendation/framework/context.go]: suggest a file name?

File: pkg/recommendation/framework/context.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+//func (ctx *RecommendationContext) Canceled() bool {
+//	select {
+//	case <-ctx.CancelCh:
+//		return true
+//	default:
+//		return false
+//	}
+//}
+
+func ObjectConversion(object interface{}, target interface{}) error {
```
[the remark this answers is not in the document]
Comment (qmhu) [reply pkg/recommendation/framework/context.go:104]: suggest a filename ?

File: pkg/recommendation/recommender/hpa/recommend.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+		}
+	}
+
+	klog.V(4).Infof("%s minMaxMedians medianMax %f, medianMin %f, medianUsages %v", rr.Name(), medianMax, medianMin, medianUsages)
+
+	return medianMin, medianMax, nil
+}
+
+// checkFluctuation check if the time series fluctuation is reach to replicas.fluctuation-threshold
+func (rr *HPARecommender) checkFluctuation(medianMin, medianMax float64) error {
```
[the remark this answers is not in the document]
Comment (qmhu) [reply pkg/recommendation/recommender/hpa/recommend.go:215]: this is a hpa recommendation function. Already remove from replicas recommender.

File: pkg/recommendation/recommender/resource/prepare.go (hunk 5 above)
Comment (qmhu) [reply pkg/recommendation/recommender/resource/prepare.go:16]: currently on.

File: pkg/recommendation/recommender/resource/recommend.go (hunk 7 above)
Comment (qmhu) [reply pkg/recommendation/recommender/resource/recommend.go:26]: every container has two algoritjm config: cpu and memory ,so cannot use context's algoritjm config. we may need to reconsider it.

**Comment by xieydd** [comment]
link: #433

**Pull request merged**